<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.305">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>GERMS-DS - Beginning Predictions</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<link href="../../about.html" rel="next">
<link href="../../docs/meetings/getting_started.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link id="quarto-text-highlighting-styles" href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../../styles/styles.scss">
</head>

<body class="docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">GERMS-DS</span>
  </a>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../../index.html" aria-current="page">Home</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../about.html">About</a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="http://germslab.org">GERMS Lab</a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Beginning Predictions</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">Home</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Meetings</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/meetings/getting_started.html" class="sidebar-item-text sidebar-link">Getting Started</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../docs/meetings/first_model.html" class="sidebar-item-text sidebar-link active">Beginning Predictions</a>
  </div>
</li>
    </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../about.html" class="sidebar-item-text sidebar-link">About</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#homework-from-last-time" id="toc-homework-from-last-time" class="nav-link active" data-scroll-target="#homework-from-last-time">Homework from last time</a>
  <ul class="collapse">
  <li><a href="#combining-the-datasets" id="toc-combining-the-datasets" class="nav-link" data-scroll-target="#combining-the-datasets">Combining the datasets</a>
  <ul class="collapse">
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data">2021 Data</a></li>
  <li><a href="#data-1" id="toc-data-1" class="nav-link" data-scroll-target="#data-1">2018 Data</a></li>
  <li><a href="#data-2" id="toc-data-2" class="nav-link" data-scroll-target="#data-2">2019 Data</a></li>
  <li><a href="#data-3" id="toc-data-3" class="nav-link" data-scroll-target="#data-3">2020 Data</a></li>
  <li><a href="#putting-it-all-together" id="toc-putting-it-all-together" class="nav-link" data-scroll-target="#putting-it-all-together">Putting it all together</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#some-brief-eda" id="toc-some-brief-eda" class="nav-link" data-scroll-target="#some-brief-eda">Some brief EDA</a>
  <ul class="collapse">
  <li><a href="#how-many-hazardous-cases-were-there-per-year" id="toc-how-many-hazardous-cases-were-there-per-year" class="nav-link" data-scroll-target="#how-many-hazardous-cases-were-there-per-year">How many hazardous cases were there per year?</a></li>
  <li><a href="#which-locations-had-the-most-harmful-algal-blooms" id="toc-which-locations-had-the-most-harmful-algal-blooms" class="nav-link" data-scroll-target="#which-locations-had-the-most-harmful-algal-blooms">Which locations had the most harmful algal blooms?</a></li>
  </ul></li>
  <li><a href="#intro-to-tidymodels" id="toc-intro-to-tidymodels" class="nav-link" data-scroll-target="#intro-to-tidymodels">Intro to Tidymodels</a>
  <ul class="collapse">
  <li><a href="#splitting-the-data-preprocessing" id="toc-splitting-the-data-preprocessing" class="nav-link" data-scroll-target="#splitting-the-data-preprocessing">Splitting the data, preprocessing</a></li>
  <li><a href="#creating-and-training-a-model" id="toc-creating-and-training-a-model" class="nav-link" data-scroll-target="#creating-and-training-a-model">Creating and training a model</a></li>
  </ul></li>
  <li><a href="#homework" id="toc-homework" class="nav-link" data-scroll-target="#homework">Homework</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Beginning Predictions</h1>
</div>





<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">4/25/22</p>
    </div>
  </div>
    
  </div>
  

</header>

<p>In this meeting, we’ll train our first model and see how to gather metrics and predictions. First, let’s go over the homework from last time.</p>
<section id="homework-from-last-time" class="level1">
<h1>Homework from last time</h1>
<p>There were three “homework questions” from last time:</p>
<ul>
<li>Repeat the EDA and visualizations for the 2019 data set.</li>
<li>Commit and push your changes</li>
<li>Combine the 2019 and 2020 datasets.</li>
</ul>
<p>The first two are pretty straightforward, so I’ll cover how to combine the 2018-2021 datasets. It’s important that we combine the datasets correctly as this dataset is what we’ll be using going forward in these meetings.</p>
<section id="combining-the-datasets" class="level2">
<h2 class="anchored" data-anchor-id="combining-the-datasets">Combining the datasets</h2>
<p>Before we get started, let’s check out the other spreadsheets. We already saw what was in the 2018 data, but let’s open the Excel spreadsheets for the other datasets to see how straightforward combining the datasets will be.</p>
<p>After opening up the datasets, here are the primary issues facing us:</p>
<ul>
<li><em>Columns have different names between the different datasets.</em> For example, the Microcystin measurement is labeled “Microcystin RAW Value [ug/L]” in the 2018 data set but is called “Microcystin” in all of the other data sets. Another example is that the gene copy numbers of the various <em>mcy</em>A species (<em>mcy</em>A P/M/A) are labeled a few different ways.</li>
<li><em>Not all columns are shared between datasets.</em> This is most obvious when comparing the 2021 dataset versus the other years.</li>
</ul>
<p>With these in mind, our plan of attack will be, for each data set:</p>
<ol type="1">
<li>Select only those columns that appear in the 2021 dataset</li>
<li>Convert column names to a standard set of names</li>
<li>Add a year column</li>
</ol>
<p>After we do all of these steps, we’ll concatenate all of the datasets together.</p>
<p>Note that this is just one way to handle this problem. Another solution might be to use <em>all</em> of the columns for all years and impute the missing data for the missing columns, or to select a learning algorithm that can handle the missing data. However, doing so introduces more uncertainty into the model we’ll use down the line, and so this particular method was chosen because…</p>
<section id="data" class="level3">
<h3 class="anchored" data-anchor-id="data">2021 Data</h3>
<p>Anyway, let’s start by reading in the 2021 data since we’ll only be using columns found for that year. After reading in the data, we’ll call <code>clean_names</code> and filter out any samples without a <code>collected_date</code>.</p>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(janitor)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2021</span> <span class="ot">&lt;-</span> <span class="fu">read_xlsx</span>(<span class="st">"data/IowaDNR_2021_Data_Merged.xlsx"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">clean_names</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(collected_date))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2021</span> <span class="sc">%&gt;%</span> </span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code> [1] "label"                  "collected_date"         "environmental_location"
 [4] "microcystin"            "tkp"                    "ortho_p"               
 [7] "tkn"                    "p_h"                    "dissolved_oxygen_mg_l" 
[10] "x16s"                   "am"                     "ap"                    
[13] "aa"                    </code></pre>
</div>
</div>
<p>Most of these columns look fine, but let’s change the <em>mcy</em> variable names from <code>ax</code> to <code>mcy_a_x</code> to be more descriptive (with the added benefit of matching how <code>clean_names</code> parses these columns in the other datasets). There are a couple ways we can do this, but I’ll be using the rename function here:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2021</span> <span class="ot">&lt;-</span> dnr<span class="fl">.2021</span> <span class="sc">%&gt;%</span> </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">mcy_a_m =</span> am,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">mcy_a_p =</span> ap,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">mcy_a_a =</span> aa</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2021</span> <span class="sc">%&gt;%</span> </span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code> [1] "label"                  "collected_date"         "environmental_location"
 [4] "microcystin"            "tkp"                    "ortho_p"               
 [7] "tkn"                    "p_h"                    "dissolved_oxygen_mg_l" 
[10] "x16s"                   "mcy_a_m"                "mcy_a_p"               
[13] "mcy_a_a"               </code></pre>
</div>
</div>
<p>Finally, we need to take care of is type of <code>pH</code>. Because there are row values of <code>No Data</code> in the spreadsheet it comes from, the <code>p_h</code> column gets read in as a character variable. Let’s mutate this column to using <code>as.numeric</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2021</span> <span class="ot">&lt;-</span> dnr<span class="fl">.2021</span> <span class="sc">%&gt;%</span> </span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">p_h =</span> <span class="fu">as.numeric</span>(p_h)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stderr">
<pre><code>Warning in mask$eval_all_mutate(quo): NAs introduced by coercion</code></pre>
</div>
</div>
<p>Great! We now know what we want the names of our columns to be and we’re ready to repeat this process for the other data sets. Let’s revisit the 2018 data.</p>
</section>
<section id="data-1" class="level3">
<h3 class="anchored" data-anchor-id="data-1">2018 Data</h3>
<p>Let’s start by looking at our column names:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2018</span> <span class="ot">&lt;-</span> <span class="fu">read_xlsx</span>(<span class="st">"data/IowaDNR_2018_Data_Merged.xlsx"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">clean_names</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(collected_date))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2018</span> <span class="sc">%&gt;%</span> </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code> [1] "sample_id"                         "environmental_location"           
 [3] "collected_date"                    "microcystin_raw_value_ug_l"       
 [5] "p_h"                               "doc_ppm"                          
 [7] "tkp_mg_p_l"                        "tkn_mg_n_l"                       
 [9] "nh3_mg_n_l"                        "n_ox_mg_n_l"                      
[11] "no2_mg_n_l"                        "cl_mg_cl_l"                       
[13] "so4_mg_so4_l"                      "x16s_r_rna_gene_copies_m_l"       
[15] "microcystismcy_a_gene_copies_m_l"  "aanabaenamcy_a_gene_copies_m_l"   
[17] "planktothrixmcy_a_gene_copies_m_l"</code></pre>
</div>
</div>
<p>The ideal situation is that all the column names match exactly and we can just use <code>any_of</code> to get those columns from this data frame. However, when we try that, we see that we still have some work to do:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2018</span> <span class="sc">%&gt;%</span> </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">any_of</span>(<span class="fu">colnames</span>(dnr<span class="fl">.2021</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code># A tibble: 539 × 3
   collected_date      environmental_location   p_h
   &lt;dttm&gt;              &lt;chr&gt;                  &lt;dbl&gt;
 1 2018-05-22 11:19:00 Backbone Beach           8.1
 2 2018-06-19 09:00:00 Backbone Beach           8.8
 3 2018-08-21 13:00:00 Backbone Beach           8.9
 4 2018-07-10 11:15:00 Backbone Beach           8.2
 5 2018-08-28 12:30:00 Backbone Beach           8.8
 6 2018-08-14 10:30:00 Backbone Beach           8  
 7 2018-08-07 11:10:00 Backbone Beach           7.7
 8 2018-07-24 12:00:00 Backbone Beach           8.9
 9 2018-07-31 11:25:00 Backbone Beach           8.5
10 2018-07-02 11:15:00 Backbone Beach           7.9
# … with 529 more rows</code></pre>
</div>
</div>
<p>We’re missing a lot of columns here. Let’s do another <code>rename</code> step here before our select step:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2018</span> <span class="ot">&lt;-</span> dnr<span class="fl">.2018</span> <span class="sc">%&gt;%</span> </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">microcystin =</span> microcystin_raw_value_ug_l,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tkp =</span> tkp_mg_p_l,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">tkn =</span> tkn_mg_n_l,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">x16s =</span> x16s_r_rna_gene_copies_m_l,</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">mcy_a_m =</span> microcystismcy_a_gene_copies_m_l,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">mcy_a_p =</span> planktothrixmcy_a_gene_copies_m_l,</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">mcy_a_a =</span> aanabaenamcy_a_gene_copies_m_l</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">any_of</span>(<span class="fu">colnames</span>(dnr<span class="fl">.2021</span>)))</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2018</span> <span class="sc">%&gt;%</span> </span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code># A tibble: 6 × 10
  collected_date      environmental_locati… microcystin   tkp   tkn   p_h   x16s
  &lt;dttm&gt;              &lt;chr&gt;                       &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
1 2018-05-22 11:19:00 Backbone Beach              0.63   1.61 0.66    8.1 1.47e5
2 2018-06-19 09:00:00 Backbone Beach              0.46   1.06 0.517   8.8 4.23e2
3 2018-08-21 13:00:00 Backbone Beach              0.438  1.06 0.169   8.9 4.04e7
4 2018-07-10 11:15:00 Backbone Beach              0.425  1.22 0.505   8.2 1.08e7
5 2018-08-28 12:30:00 Backbone Beach              0.418  1.02 1.14    8.8 7.65e6
6 2018-08-14 10:30:00 Backbone Beach              0.405  1.15 0.298   8   4.46e7
# … with 3 more variables: mcy_a_m &lt;dbl&gt;, mcy_a_p &lt;dbl&gt;, mcy_a_a &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Much better. Note that there are some missing columns here because there are columns in the 2021 dataset that don’t exist in the 2018 dataset.</p>
</section>
<section id="data-2" class="level3">
<h3 class="anchored" data-anchor-id="data-2">2019 Data</h3>
<p>We’ll repeat this process for the 2019 dataset. Some notes:</p>
<ul>
<li><code>clean_names</code> already nicely formatted the <em>mcy</em>X columns</li>
<li>By default, <code>read_xlsx</code> reads in the first sheet of the <code>xlsx</code> file. This means that the 2019 data will only show the data on the first sheet. We want the data from the sheet labeled “combined”, so use the <code>sheet</code> argument of <code>read_xlsx</code> to do this.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2019</span> <span class="ot">&lt;-</span> <span class="fu">read_xlsx</span>(<span class="st">"data/IowaDNR_2019_Data_Merged.xlsx"</span>,</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">sheet =</span> <span class="st">"combined"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">clean_names</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(collected_date)) <span class="sc">%&gt;%</span> </span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">tkn =</span> tkn_mg_n_l,</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">tkp =</span> tkp_mg_p_l,</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">ortho_p =</span> ortho_p_mg_p_l</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">any_of</span>(<span class="fu">colnames</span>(dnr<span class="fl">.2021</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2019</span> <span class="sc">%&gt;%</span> </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code># A tibble: 6 × 12
  label    collected_date      environmental_l… microcystin    tkp ortho_p   tkn
  &lt;chr&gt;    &lt;dttm&gt;              &lt;chr&gt;                  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
1 6-21280… 2019-06-25 00:00:00 Backbone Beach         0      0.473  0.065  0.18 
2 7-21280… 2019-07-01 00:00:00 Backbone Beach         0      0.023  0.034  0.282
3 8-21280… 2019-07-09 00:00:00 Backbone Beach         0.23   0.056  0.0401 0.346
4 9-21280… 2019-07-16 00:00:00 Backbone Beach         0.743 -0.005  0.023  0.365
5 10-2128… 2019-07-23 00:00:00 Backbone Beach         0.038  0.287  0.081  0.858
6 11-2128… 2019-07-30 00:00:00 Backbone Beach         0.105  0.199  0.22   1.18 
# … with 5 more variables: p_h &lt;dbl&gt;, x16s &lt;dbl&gt;, mcy_a_m &lt;dbl&gt;, mcy_a_p &lt;dbl&gt;,
#   mcy_a_a &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="data-3" class="level3">
<h3 class="anchored" data-anchor-id="data-3">2020 Data</h3>
<p>Pretty much the same process as above, though we need to cast the <code>ortho_p</code>, <code>tkn</code>, and <code>tkp</code> columns as a numeric to account for places where the text value “NA” is entered:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>dnr<span class="fl">.2020</span> <span class="ot">&lt;-</span> <span class="fu">read_xlsx</span>(<span class="st">"data/IowaDNR_2020_Data_Merged.xlsx"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">clean_names</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(collected_date)) <span class="sc">%&gt;%</span> </span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">tkp =</span> tkp_mg_p_l,</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">tkn =</span> tkn_mg_n_l, </span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">ortho_p =</span> ortho_p_mg_p_l</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">across</span>(<span class="fu">c</span>(ortho_p, tkn, tkp), <span class="sc">~</span> <span class="fu">as.numeric</span>(.))</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">any_of</span>(<span class="fu">colnames</span>(dnr<span class="fl">.2021</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-stderr">
<pre><code>Warning in mask$eval_all_mutate(quo): NAs introduced by coercion

Warning in mask$eval_all_mutate(quo): NAs introduced by coercion</code></pre>
</div>
</div>
</section>
<section id="putting-it-all-together" class="level3">
<h3 class="anchored" data-anchor-id="putting-it-all-together">Putting it all together</h3>
<p>Since we did all of the cleaning ahead of time, combining the columns now will be straightforward using <code>bind_rows</code>. Let’s save this to a new variable called <code>dnr_all</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>dnr.all <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  dnr<span class="fl">.2018</span>,</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  dnr<span class="fl">.2019</span>,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  dnr<span class="fl">.2020</span>,</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  dnr<span class="fl">.2021</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target="#callout-1" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Should we have written functions?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Notice that there were a lot of redundant steps across all of the DNR sets. For each data set, we:</p>
<ol type="1">
<li>Read in the Excel spreadsheet</li>
<li>Called <code>clean_names</code></li>
<li>Filtered <code>na</code> dates</li>
<li>Renamed some columns</li>
<li>Selected the columns that match the names of the 2021 dataset.</li>
</ol>
<p>Ideally, we would have written a function to handle all of the redundant tasks. For example, we could have written a function <code>read_dnr_sheet</code> that takes in a file path, reads in the Excel sheet, calls <code>clean_names</code>, filters <code>na</code> dates, renames columns, and selects only those columns that match those from the 2021 data.</p>
<p>So yes, it would have been nice to write functions. However, the problem is that the data sheets are all just different enough that this isn’t exactly quite worth the effort. For instance, the 2019 data requires the <code>sheet</code> argument, which complicates the potential function definition. Similarly, the column names that are required to be renamed don’t quite match up between the different years, and so would require an additional argument.</p>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="some-brief-eda" class="level1">
<h1>Some brief EDA</h1>
<p>Before we jump into predictions, let’s do some quick visualizations. First, let’s create the <code>hazard_class</code> column to record whether or not the microcystin concentration was over 8 ug/L for that observation, as well as a <code>year</code> column so we can separate the samples by the year they were collected in. Finally, we’ll remove observations that don’t have a valid hazard class:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lubridate)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>dnr.all <span class="ot">&lt;-</span> dnr.all <span class="sc">%&gt;%</span> </span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">hazard_class =</span> <span class="fu">if_else</span>(microcystin <span class="sc">&gt;</span> <span class="dv">8</span>, <span class="st">"hazardous"</span>,<span class="st">"safe"</span>),</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">year =</span> <span class="fu">year</span>(collected_date),</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">dissolved_oxygen_mg_l =</span> <span class="fu">as.numeric</span>(dissolved_oxygen_mg_l)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span> </span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(hazard_class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="how-many-hazardous-cases-were-there-per-year" class="level2">
<h2 class="anchored" data-anchor-id="how-many-hazardous-cases-were-there-per-year">How many hazardous cases were there per year?</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_light</span>())</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>dnr.all <span class="sc">%&gt;%</span> </span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(hazard_class <span class="sc">==</span> <span class="st">"hazardous"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(year) <span class="sc">%&gt;%</span> </span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(year, n, <span class="at">fill =</span> n)) <span class="sc">+</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span> </span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>(<span class="at">option =</span> <span class="st">"inferno"</span>, <span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">end =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Sample year"</span>,</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"# HABs Observed"</span>,</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"# Harmful algal blooms observed per year"</span>,</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill =</span> <span class="st">"# HABs</span><span class="sc">\n</span><span class="st">observed"</span></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> </span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.grid.major.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.line.x =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="fl">0.5</span>),</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.line.y =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="fl">0.5</span>),</span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>()</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> </span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(</span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.1</span>), <span class="at">add =</span> <span class="dv">0</span>)</span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="first_model_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="which-locations-had-the-most-harmful-algal-blooms" class="level2">
<h2 class="anchored" data-anchor-id="which-locations-had-the-most-harmful-algal-blooms">Which locations had the most harmful algal blooms?</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>dnr.all <span class="sc">%&gt;%</span> </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(hazard_class <span class="sc">==</span> <span class="st">"hazardous"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(environmental_location) <span class="sc">%&gt;%</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">environmental_location =</span> <span class="fu">fct_reorder</span>(environmental_location, n)) <span class="sc">%&gt;%</span> </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(n, environmental_location, <span class="at">fill =</span> n)) <span class="sc">+</span> </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span> </span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"# HABS observed (all years)"</span>,</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Sample Location"</span>,</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Total number of HABs observed for all years"</span>,</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill =</span> <span class="st">"# HABs</span><span class="sc">\n</span><span class="st">Observed"</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> </span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_c</span>(<span class="at">option =</span> <span class="st">"magma"</span>, <span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.grid.major.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>()</span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> </span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(</span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.1</span>), <span class="at">add =</span> <span class="dv">0</span>)</span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="first_model_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="intro-to-tidymodels" class="level1">
<h1>Intro to Tidymodels</h1>
<p>Up until now, we’ve just been doing EDA stuff. Now that we’ve done some of that, we can get started with modeling. For this series, we’re going to use <code>tidymodels</code> to keep our prediction workflows organized and reproducible.</p>
<p>The <a href="https://www.tidymodels.org/"><code>tidymodels</code></a> framework is “a collection of packages for modeling and machine learning using tidyverse principles.” The <code>tidymodels</code> website has <a href="https://www.tidymodels.org/start/">several tutorials</a> to introduce you to the fundamentals of working with <code>tidymodels</code>. Today, we’re going to use the <code>tidymodels</code> framework to:</p>
<ol type="1">
<li>Split our data</li>
<li>Perform some preprocessing</li>
<li>Train a linear model</li>
<li>Visualize some metrics</li>
</ol>
<section id="splitting-the-data-preprocessing" class="level2">
<h2 class="anchored" data-anchor-id="splitting-the-data-preprocessing">Splitting the data, preprocessing</h2>
<p>We begin by splitting our data into testing and training splits:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidymodels)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">489</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>data_split <span class="ot">&lt;-</span> <span class="fu">initial_split</span>(dnr.all, <span class="at">strata =</span> <span class="st">"hazard_class"</span>)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>training_data <span class="ot">&lt;-</span> <span class="fu">training</span>(data_split)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>testing_data <span class="ot">&lt;-</span> <span class="fu">testing</span>(data_split)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we create a recipe. A recipe is a series of data processing steps to perform on our data before we feed it into our model. Here, the only steps we’ll take is to remove non-informative variables and to impute some missing values by filling in the mean.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>hab_recipe <span class="ot">&lt;-</span> <span class="fu">recipe</span>(microcystin <span class="sc">~</span> ., training_data) <span class="sc">%&gt;%</span> </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_rm</span>(collected_date, label) <span class="sc">%&gt;%</span> </span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_impute_mean</span>(<span class="fu">all_numeric</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can take a look at the preprocessed data set by using the <code>prep</code> and <code>juice</code> functions. <code>prep</code> will perform the steps outlined in the pre-processing recipe. For example, in our recipe above, it will calculate the mean of all the numeric columns. <code>juice</code> will then return the transformed dataset resulting from following all the steps in the recipe:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>hab_recipe <span class="sc">%&gt;%</span> </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  prep <span class="sc">%&gt;%</span> </span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  juice</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code># A tibble: 1,193 × 13
   environmental_loca…   tkp    tkn   p_h   x16s mcy_a_m mcy_a_p mcy_a_a ortho_p
   &lt;fct&gt;               &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 Backbone Beach      1.61   0.66   8.1  1.47e5       0       0       0  0.0430
 2 Backbone Beach      1.06   0.517  8.8  4.23e2       0       0       0  0.0430
 3 Backbone Beach      1.22   0.505  8.2  1.08e7    1240       0       0  0.0430
 4 Backbone Beach      1.02   1.14   8.8  7.65e6       0       0       0  0.0430
 5 Backbone Beach      1.15   0.298  8    4.46e7       0       0       0  0.0430
 6 Backbone Beach      0.922 -0.24   7.7  2.66e7       0       0       0  0.0430
 7 Backbone Beach      0.041  0.233  8.9  5.04e6       0       0       0  0.0430
 8 Backbone Beach      0.491  0.264  8.5  1.09e7     432       0       0  0.0430
 9 Backbone Beach      1.12   0.437  7.9  6.09e6       0       0       0  0.0430
10 Backbone Beach      1.22   0.569  7.28 4.44e6       0       0       0  0.0430
# … with 1,183 more rows, and 4 more variables: dissolved_oxygen_mg_l &lt;dbl&gt;,
#   hazard_class &lt;fct&gt;, year &lt;dbl&gt;, microcystin &lt;dbl&gt;</code></pre>
</div>
</div>
<p>We see that the <code>collected_date</code> and <code>label</code> were removed, and the numeric columns had the missing data filled in with the column mean.</p>
</section>
<section id="creating-and-training-a-model" class="level2">
<h2 class="anchored" data-anchor-id="creating-and-training-a-model">Creating and training a model</h2>
<p>Now that we have our data preprocessing steps, we’ll create the model that we want to use to predict. To begin with, we’ll just use an <code>lm</code> model. We’ll indicate this to <code>tidymodels</code> using the <code>linear_reg</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>linear_model <span class="ot">&lt;-</span> <span class="fu">linear_reg</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_engine</span>(<span class="st">"lm"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_mode</span>(<span class="st">"regression"</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>linear_model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code>Linear Regression Model Specification (regression)

Computational engine: lm </code></pre>
</div>
</div>
<p>This tells us that <code>linear_model</code> is a <code>tidymodels</code> object using the <code>lm</code> function to perform linear regression.</p>
<p>Now we can package the preprocessing recipe and the model into a workflow. Think of the workflow as a pipeline - a series of steps you want to carry out on the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>lm_workflow <span class="ot">&lt;-</span> <span class="fu">workflow</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_recipe</span>(hab_recipe) <span class="sc">%&gt;%</span> </span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_model</span>(linear_model) </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>lm_workflow</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code>══ Workflow ════════════════════════════════════════════════════════════════════
Preprocessor: Recipe
Model: linear_reg()

── Preprocessor ────────────────────────────────────────────────────────────────
2 Recipe Steps

• step_rm()
• step_impute_mean()

── Model ───────────────────────────────────────────────────────────────────────
Linear Regression Model Specification (regression)

Computational engine: lm </code></pre>
</div>
</div>
<p>This shows us all of the preprocessing steps we’re going to carry out on the data as well as the model we’re going to use. We’re now ready to fit our model to the training data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>lm_fit <span class="ot">&lt;-</span> lm_workflow <span class="sc">%&gt;%</span> </span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">last_fit</span>(<span class="at">split =</span> data_split)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stderr">
<pre><code>! train/test split: preprocessor 1/1, model 1/1 (predictions): prediction from a rank-defici...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>lm_fit</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code># Resampling results
# Manual resampling 
# A tibble: 1 × 6
  splits             id               .metrics .notes   .predictions .workflow 
  &lt;list&gt;             &lt;chr&gt;            &lt;list&gt;   &lt;list&gt;   &lt;list&gt;       &lt;list&gt;    
1 &lt;split [1193/398]&gt; train/test split &lt;tibble&gt; &lt;tibble&gt; &lt;tibble&gt;     &lt;workflow&gt;

There were issues with some computations:

  - Warning(s) x1: prediction from a rank-deficient fit may be misleading

Use `collect_notes(object)` for more information.</code></pre>
</div>
</div>
<p>The <code>last_fit</code> function takes the entire data split, performs the preprocessing and model training steps on the training set, then predicts on the testing set. There are several convenience functions we can use to interact with the <code>lm_fit</code> object. To start, we can collect the metrics via <code>collect_metrics</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>lm_fit <span class="sc">%&gt;%</span> </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  collect_metrics</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code># A tibble: 2 × 4
  .metric .estimator .estimate .config             
  &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt;               
1 rmse    standard       3.42  Preprocessor1_Model1
2 rsq     standard       0.692 Preprocessor1_Model1</code></pre>
</div>
</div>
<p>We can also collect the predictions via <code>collect_predictions</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>lm_fit <span class="sc">%&gt;%</span> </span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect_predictions</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-stdout">
<pre><code># A tibble: 398 × 5
   id               .pred  .row microcystin .config             
   &lt;chr&gt;            &lt;dbl&gt; &lt;int&gt;       &lt;dbl&gt; &lt;chr&gt;               
 1 train/test split 0.967     3       0.438 Preprocessor1_Model1
 2 train/test split 0.661    13       0.153 Preprocessor1_Model1
 3 train/test split 1.53     21       0.377 Preprocessor1_Model1
 4 train/test split 2.12     23       0.195 Preprocessor1_Model1
 5 train/test split 2.03     30       0     Preprocessor1_Model1
 6 train/test split 0.654    31       0.855 Preprocessor1_Model1
 7 train/test split 0.837    34       0.69  Preprocessor1_Model1
 8 train/test split 0.677    36       0.405 Preprocessor1_Model1
 9 train/test split 0.673    43       0.003 Preprocessor1_Model1
10 train/test split 0.423    44       0     Preprocessor1_Model1
# … with 388 more rows</code></pre>
</div>
</div>
<p>This is perhaps more useful when visualized:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>lm_fit <span class="sc">%&gt;%</span> </span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect_predictions</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(.pred, microcystin)) <span class="sc">+</span> </span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"#0000FF"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">aspect.ratio =</span> <span class="dv">1</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Predicted"</span>,</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Actual"</span>,</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Linear Regression Prediction Results - dnr.all dataset"</span></span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="first_model_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target="#callout-2" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Typically when working with <code>tidymodels</code>, there are a few intermediate steps before jumping straight to <code>last_fit</code> and looking at predictions. For another illustration of using <code>tidymodels</code> on the built-in <code>diamonds</code> data set, see the code below, borrowed from <a href="https://stackoverflow.com/a/63239991/6024276">this StackOverflow post</a>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">489</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>diamonds_split <span class="ot">&lt;-</span> <span class="fu">initial_split</span>(diamonds, <span class="at">prop =</span> <span class="dv">4</span><span class="sc">/</span><span class="dv">5</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>diamonds_train <span class="ot">&lt;-</span> <span class="fu">training</span>(diamonds_split)</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>diamonds_test <span class="ot">&lt;-</span> <span class="fu">testing</span>(diamonds_split)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>diamonds_recipe <span class="ot">&lt;-</span> </span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">recipe</span>(price <span class="sc">~</span> ., <span class="at">data =</span> diamonds_train) <span class="sc">%&gt;%</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_log</span>(<span class="fu">all_outcomes</span>(),<span class="at">skip =</span> T) <span class="sc">%&gt;%</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_normalize</span>(<span class="fu">all_predictors</span>(), <span class="sc">-</span><span class="fu">all_nominal</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_dummy</span>(<span class="fu">all_nominal</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">step_poly</span>(carat, <span class="at">degree =</span> <span class="dv">2</span>)</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>lr_model <span class="ot">&lt;-</span> </span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">linear_reg</span>()<span class="sc">%&gt;%</span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_engine</span>(<span class="st">"lm"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_mode</span>(<span class="st">"regression"</span>)</span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a>lr_workflow <span class="ot">&lt;-</span> <span class="fu">workflow</span>() <span class="sc">%&gt;%</span></span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_model</span>(lr_model) <span class="sc">%&gt;%</span> </span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_recipe</span>(diamonds_recipe) </span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a>final_model <span class="ot">&lt;-</span> <span class="fu">fit</span>(lr_workflow, diamonds)</span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a><span class="fu">predict</span>(final_model, <span class="at">new_data =</span> <span class="fu">as.data.frame</span>(diamonds_test))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="homework" class="level1">
<h1>Homework</h1>
<ul>
<li>So far, we’ve been doing all of our work within one <code>qmd</code>. However, the code to read in the data and combine it is getting quite big. Let’s simplify this document by moving our data prep (reading in the data sets, cleaning the data, combining the data) into a R script file called <code>load_DNR_data.R</code>. After creating <code>load_DNR_data.R</code>, <code>source</code> the script at the top of your <code>qmd</code>.</li>
<li>Recreate the plot below:</li>
</ul>
<div class="cell">
<div class="cell-output-display">
<p><img src="first_model_files/figure-html/unnamed-chunk-60-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Here are some hints:</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target="#callout-3" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Filling in the missing values
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>There are a couple different ways to fill in the missing values. The way I achived this was by using <code>pivot_wider</code>, <code>replace</code>, and <code>pivot_longer</code>. Another potential way is to use <code>join/anti_join</code>.</p>
</div>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target="#callout-4" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
The fonts
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The font used in this visualization is <a href="https://fonts.google.com/specimen/Patua+One">Patua One</a>. I used the <code>showtext</code> and <code>ggtext</code> packages in order to use the font. Specifically, the <code>font_add_google</code>, <code>showtext_auto</code>, and <code>element_textbox</code> functions were used.</p>
</div>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target="#callout-5" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
The highlights in the title
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The <code>glue</code> package was used to construct the title. The words to be highlighted were surrounded by <code>span</code> elements to change the color. The color used was <code>#0000FF</code>.</p>
</div>
</div>
</div>
<ul>
<li>Evaluate the performance of the linear regression on the prediction of HAB occurrence.</li>
<li>There was a warning when we created the <code>lm_fit</code> object. What was the warning? What does that mean?
<ul>
<li>Hint: Check out <code>lm_fit$.workflow[[1]] %&gt;% tidy()</code></li>
</ul></li>
<li>What are some problems you can see with the current state of <code>dnr.all</code>?</li>
<li>What are some issues you see with the correct predictions output by <code>lm_fit</code>?</li>
<li>Add another preprocessing step to the model recipe. For example, add a step to normalize the data or another step to turn <code>environmental_location</code> into a dummy variable.</li>
<li>Install the <code>xgboost</code> library and change the model used to <code>xgboost</code>. Between classification and regression, which models performed better? Why?
<ul>
<li>Hint: Use <code>boost_tree</code> and <code>set_mode("classification")</code></li>
</ul></li>
<li>Install the <code>glmnet</code> library and use a logistic regression model. Between the XGBoost and logistic regression models, which performed better?</li>
<li>Our data is super imbalanced. How should we handle this?</li>
<li>Look at summary of <code>lm_fit</code> again. What red flags are you seeing?</li>
<li>What additional steps should we be doing in our model training workflow?</li>
</ul>


</section>

</main> <!-- /main -->
<script type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../docs/meetings/getting_started.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Getting Started</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../about.html" class="pagination-link">
        <span class="nav-page-text">About</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>